### PR TITLE
[NewsletterAdapter] change to use configuration and container

### PR DESCRIPTION
**Pull request by dpfaffenbauer**
This relates to issue #1952

It now uses a configuration to register source adapters and a service-locator to retrieve them in controller/command.

```yml
pimcore:
  documents:
    newsletter:
      adapters:
         identifier: service-name
```

Also the AddressSourceAdapterInterface has been changed: renamed __construct to configure to be able to inject services.

### CODE REVIEW

**Comment by dpfaffenbauer** [comment]
Oh, just saw: PimcoreCoreExtension changed a bit more than intented cause of automated Code Formatting from PHP-Storm, hope you don't mind?

**Comment by dpfaffenbauer** [comment]
Still not quite happy with the proposed solution. Actually, the ```configure``` function should be inside a factory and prepare the SourceAdapter, the SourceAdapter then gets passed around. so it wouldn't be stateful as it is now.

But, to do that, we would need to introduce a new AddressSourceFactory and add a factory for each type. What do you think about that?

**Comment by maff** [comment]
@dpfaffenbauer thanks for the PR :) I'm afraid the stateful solution imposes some unforseeable risks like state not being reinitialized on configure (e.g. https://github.com/pimcore/pimcore/blob/master/pimcore/lib/Pimcore/Document/Newsletter/AddressSourceAdapter/DefaultAdapter.php#L66). Although this could be handled by implementations (e.g. resetting state on `configure()`) I think a solution building a new adapter is more favorable. I'd suggest:

* Build adapters through a factory as you suggested
* Implement a configurable default factory which can be configured with a class name to build and just passes the parameters is receives to the adapter constructor (maybe something like [AgentFactory](https://github.com/pimcore/pimcore/blob/master/pimcore/lib/Pimcore/Bundle/EcommerceFrameworkBundle/OrderManager/Order/AgentFactory.php)). Register 3 factory services using this factory class and update the config mapping to point to the factory service instead of the adapter service (which will be removed). More complex adapter implementations can be built by custom factories. 
* Build a locator as your did for the adapters, but containing the factories this time.

WDYT?

**Comment by dpfaffenbauer** [comment]
@maff yeah, thought so too 👍 

Implemented using Factory Pattern now.

File: doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/02_V4_to_V5.md [hunk 1]
```diff
@@ -296,3 +296,23 @@ $this->get(\Pimcore\Cache\Runtime::class) // service container
...
+      class: Pimcore\Document\Newsletter\DefaultAddressSourceAdapterFactory
+      arguments:
+        - 'App\Document\Newsletter\AddressSourceAdapter\MyAdapter'
+  ```
+- If you are using a more complex Source Adapter, you can create your own Factory by implementing the interface Pimcore\Document\Newsletter\Factory\AddressSourceAdapterFactoryInterface
+- Add your Adapter Factory to the configuration:
+
+```yml
+pimcore:
+    newsletter:
```
Comment (maff) [inline doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/02_V4_to_V5.md]: It's currently `pimcore.documents.newsletter` but for me `pimcore.newsletter` sounds better/more concise. Could you update the config + extension accordingly?

File: doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/02_V4_to_V5.md [hunk 2]
```diff
@@ -296,3 +296,23 @@ $this->get(\Pimcore\Cache\Runtime::class) // service container
...
+        adapters:
```
Comment (maff) [inline doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/02_V4_to_V5.md]: `source_adapters` maybe?

File: pimcore/lib/Pimcore/Bundle/CoreBundle/Command/InternalNewsletterDocumentSendCommand.php [hunk 3]
```diff
@@ -63,12 +64,19 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $addressSourceAdapterName = $data['addressSourceAdapterName'];
         $adapterParams = $data['adapterParams'];
 
-        $adapterClass = '\\Pimcore\\Document\\Newsletter\\AddressSourceAdapter\\' . ucfirst($addressSourceAdapterName);
+        $serviceLocator = $this->getContainer()->get('pimcore.newsletter.address_source_adapter.factories');
+
+        if (!$serviceLocator->has($addressSourceAdapterName)) {
+            //TODO: Shouldn't this throw an exception?
```
Comment (maff) [inline pimcore/lib/Pimcore/Bundle/CoreBundle/Command/InternalNewsletterDocumentSendCommand.php]: I would definitely throw an exception here. Previously a class not found exception was thrown, now we just can use a custom one.

File: pimcore/lib/Pimcore/Bundle/CoreBundle/Resources/config/pimcore/config.yml [hunk 4]
```diff
@@ -150,6 +150,11 @@ pimcore:
             prefixes:
                 - \Pimcore\Model\Document\Tag\
                 - \Document_Tag_
+        newsletter:
+            adapters:
+              defaultAdapter: pimcore.document.newsletter.factory.default
```
Comment (maff) [inline pimcore/lib/Pimcore/Bundle/CoreBundle/Resources/config/pimcore/config.yml]: indented by 4 please :)

File: pimcore/lib/Pimcore/Document/Newsletter/AddressSourceAdapterFactoryInterface.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+interface AddressSourceAdapterFactoryInterface
+{
+    /**
+     * Configures and creates the AddressSourceAdapterInterface
+     *
+     * @param $params
+     * @return AddressSourceAdapterInterface
+     */
+    public function configure($params);
```
Comment (maff) [inline pimcore/lib/Pimcore/Document/Newsletter/AddressSourceAdapterFactoryInterface.php]: I'd prefer a name which indicates that an instance is built here (create, build, ...). `configure()` sounds more like "take an existing instance and add configuration".

**Comment by dpfaffenbauer** [comment]
@maff all requested changes done

File: pimcore/lib/Pimcore/Bundle/CoreBundle/Command/InternalNewsletterDocumentSendCommand.php [hunk 6]
```diff
@@ -16,11 +16,13 @@
 
 use Pimcore\Console\AbstractCommand;
 use Pimcore\Document\Newsletter\AddressSourceAdapterInterface;
+use Pimcore\Document\Newsletter\AddressSourceAdapterFactoryInterface;
 use Pimcore\Logger;
 use Pimcore\Model;
 use Pimcore\Tool\Newsletter;
 use Symfony\Component\Console\Input\InputInterface;
 use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Intl\Exception\RuntimeException;
```
Comment (maff) [inline pimcore/lib/Pimcore/Bundle/CoreBundle/Command/InternalNewsletterDocumentSendCommand.php]: This was accidentally i suppose?

File: doc/Development_Documentation/03_Documents/02_Document_Types/05_Newsletter_Documents.md [hunk 7]
```diff
@@ -46,9 +46,25 @@ It is easily possible to implement custom address source adapter. Following file
...
+the email addresses. It needs to implement the interface `AddressSourceAdapterInterface`. As sample see
+ [csvList](https://github.com/pimcore/pimcore/blob/master/pimcore/lib/Pimcore/Document/Newsletter/AddressSourceAdapter/CsvList.php).
+- Register your Adapter Source Factory as Service. If you are using a very simple Adapter Source, you can use the DefaultFactory
+  ```yml
+  app.document.newsletter.factory.myAdapter:
+      class: Pimcore\Document\Newsletter\DefaultAddressSourceAdapterFactory
+      arguments:
+        - 'App\Document\Newsletter\AddressSourceAdapter\MyAdapter'
+  ```
+- If you are using a more complex Source Adapter, you can create your own Factory by implementing the interface Pimcore\Document\Newsletter\Factory\AddressSourceAdapterFactoryInterface
```
Comment (maff) [inline doc/Development_Documentation/03_Documents/02_Document_Types/05_Newsletter_Documents.md]: `Pimcore\Document\Newsletter\AddressSourceAdapterFactoryInterface`

File: pimcore/lib/Pimcore/Bundle/CoreBundle/Command/InternalNewsletterDocumentSendCommand.php (hunk 6 above)
Comment (dpfaffenbauer) [reply pimcore/lib/Pimcore/Bundle/CoreBundle/Command/InternalNewsletterDocumentSendCommand.php]: no, thats the exception thrown in line 71. Better idea for a different Exception Type? Maybe ClassNotFound or InvalidArgument?

Comment (maff) [reply pimcore/lib/Pimcore/Bundle/CoreBundle/Command/InternalNewsletterDocumentSendCommand.php]: `\RuntimeException` or `\InvalidArgumentException` but in any case none of Symfony's intl component.

**Comment by dpfaffenbauer** [comment]
alright, resolved again :)

**Comment by maff** [comment]
Thanks a lot :)

**Comment by dpfaffenbauer** [comment]
awesome :) 

🍻

**Comment by dpfaffenbauer** [comment]
@maff Just encountered an issuse in this line: https://github.com/pimcore/pimcore/blob/master/pimcore/lib/Pimcore/Bundle/CoreBundle/DependencyInjection/PimcoreCoreExtension.php#L405

Since the service may not be available at the point of time the service-locator gets defined, this will not work as expected. Solution: Just remove the exception and it should run into runtime issues when service is not available, what do you think?

**Comment by maff** [comment]
@dpfaffenbauer you're right, that check is too early. However, Symfony checks for invalid references in `CheckExceptionOnInvalidReferenceBehaviorPass` so we can just remove that check and rely on Symfony's handling :) Would you provide a PR?

**Comment by dpfaffenbauer** [comment]
alright, PR is coming.

Could you do me another favor pls? Could someone please review PR 1880 before stable release?

**Pull request merged**